### DISCUSSION TITLE
Update the list of approved users

### DISCUSSION THREAD

**Discussion by slawiko** [umputun/tg-spam#13] [Q&A] [answered]
It is not clear how to update the list of approved users. 

It only loads list of users from file during the bot start: https://github.com/umputun/tg-spam/blob/master/app/main.go#L162 , but how (and when?) can I add user to that list.

My assumption is that list has to be updated after clicking unban admin?

Anyway, could you please explain how it should work from your perspective and I will do my best to fix that if needed.

**Comment by umputun** [comment]
This is fully automatic; the user can't change it manually. Please see docs:

> Another useful feature is the ability to keep the list of approved users persistently. The bot will not ban those users and won't check their messages for spam because they have already passed the initial check. IDs of those users are kept in the internal list and stored in the file approved-users.txt. To enable this feature, the user must specify the file with the list of approved users with --files.approved-users=, [$FILES_APPROVED_USERS] parameter. The file is binary and can't be edited manually. The bot handles it automatically if the parameter is set and --paranoid mode is not enabled. 

_i'll move it to dicusssions_

**Reply by slawiko** [reply]
But how is the approved users list updated? Where is it happening and under which circumstances?

My assumption was that approved users list should be updated by manually unbanning users through the console.

**Reply by umputun** [reply, marked as answer]
if `--paranoid` mode is not enabled a user will be checked once, on the first message. As long as the message is ham (not spam) the user will be added to the approved list

**Reply by slawiko** [reply]
Finally found the place: https://github.com/umputun/tg-spam/blob/v0.2.2/lib/detector.go#L123

thanks!
